SmartContract as trait

Maar-io opened the pull request:
@akru @Dinonard

please take a look at this commit.
The implementation of trait `IsContract` is placed in runtime in order to reach EVM module.
But now the visibility of fn `is_contract` in pallet-dapps-staking is moved as well and compiler complains with:
`
no variant or associated item namedis_contractfound for enumtraits::SmartContractin the current scope
`

please advice what would be the best approach to solve this

akru commented on frame/dapps-staking/src/traits.rs:
```diff
@@ -0,0 +1,14 @@
+    Wasm(AccountId),
+}
+
+pub trait IsContract<AccountId> {
```
Parameter isn’t required here

Maar-io replied:
changed

Maar-io commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -302,7 +305,7 @@ pub mod pallet {
             );
             ensure!(
-                Self::is_contract_valid(&contract_id),
+                SmartContract::is_contract(&contract_id),
```
@akru the error is similar after the proposed change
```
error[E0599]: no method named `is_contract` found for enum `traits::SmartContract` in the current scope
   --> frame/dapps-staking/src/pallet/mod.rs:307:33
```

Dinonard replied:
I think the issue is that you're still using SmartContract enum in dapps-staking pallet.

It should be moved outside of the pallet and you should only depend on IsContract trait.

E.g. method footprint should be:
        pub fn register(
            origin: OriginFor<T>,
            contract_id: IsContract,
) -> DispatchResultWithPostInfo {


Maybe an additional trait bound is needed since the idea is to use this value as 'id'.

Dinonard commented on frame/dapps-staking/src/traits.rs:
```diff
@@ -0,0 +1,14 @@
+use super::*;
+
+/// Multi-VM pointer to smart contract instance.
+#[derive(PartialEq, Eq, Copy, Clone, Encode, Decode, RuntimeDebug)]
+pub enum SmartContract<AccountId> {
+    /// EVM smart contract instance.
+    Evm(sp_core::H160),
+    /// Wasm smart contract instance.
+    Wasm(AccountId),
+}
```
You should move this to the runtime implementation.

dapps-staking pallet doesn't need to know about such details.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -301,10 +304,7 @@ pub mod pallet {
-                Self::is_contract_valid(&contract_id),
-                Error::<T>::ContractIsNotValid
-            );
+            ensure!(contract_id.is_contract(), Error::<T>::ContractIsNotValid);
```
You probably get these errors because you haven't imported the trait.

I might be wrong though.

akru reviewed (approved):
LGTM

Maar-io commented:
@akru @Dinonard please have another look. Changes:
- updated Shibuya runtime
- rebased to `feature/dapp-staking-pallet`
- changed Shibuya configurable constants
- fix bug with registrationDeposit amount
- removed some unused code and comments
- HistoryDepth set initially to be 30 days

akru commented on runtime/shibuya/src/lib.rs:
```diff
@@ -289,6 +291,24 @@ impl pallet_dapps_staking::Config for Runtime {
+    /// EVM smart contract instance.
+    Evm(sp_core::H160),
+    /// Wasm smart contract instance.
+    Wasm(AccountId),
```
Shibuya have no wasm contracts now, let’s guard users of errors

akru reviewed (changes requested):
Looks good, but please fix shibuya runtime

Dinonard commented on frame/dapps-staking/src/traits.rs:
```diff
@@ -0,0 +1,3 @@
+pub trait IsContract {
+    fn is_contract(&self) -> bool;
```
Imo, this should just be in the mod.rs of the pallet.

Maar-io replied:
there is no fast rule on where the traits should be defined. We can do it either way

Dinonard replied:
I commented since it's 3 lines of code.
You could re-export it as something from pallet_dapps_staking::trait if you want to use trait.

But it's fine like this too.

Now that I wrote the comment above, I realized there's no documentation on the method :P

Dinonard commented:
Change request resolved in Slack.
